File: backend/packages/db/src/db/repositories/inbox.py
from __future__ import annotations

from uuid import UUID

from sqlalchemy import and_, desc, func, select, update
from sqlalchemy.orm import Session

from db.models.inbox_item import InboxItem

FILTER_TO_TYPES = {
    "all": None,
    "invites": ["invite"],
    "prechats": ["prechat_summary"],
    "matches": ["match_notification"],
    "system": ["system"],
}


def list_inbox_items_for_user(session: Session, user_id: UUID, filter_name: str = "all") -> list[InboxItem]:
    statement = select(InboxItem).where(InboxItem.user_id == user_id)
    types = FILTER_TO_TYPES[filter_name]
    if types is not None:
        statement = statement.where(InboxItem.item_type.in_(types))
    statement = statement.order_by(desc(InboxItem.created_at))
    return list(session.execute(statement).scalars())


def get_inbox_item_for_user(session: Session, item_id: UUID, user_id: UUID) -> InboxItem | None:
    statement = select(InboxItem).where(and_(InboxItem.id == item_id, InboxItem.user_id == user_id)).limit(1)
    return session.execute(statement).scalar_one_or_none()


def get_unread_inbox_count(session: Session, user_id: UUID) -> int:
    statement = select(func.count()).select_from(InboxItem).where(
        and_(InboxItem.user_id == user_id, InboxItem.read.is_(False))
    )
    return int(session.execute(statement).scalar_one())


def mark_inbox_item_read(session: Session, item_id: UUID, user_id: UUID) -> bool:
    statement = (
        update(InboxItem)
        .where(and_(InboxItem.id == item_id, InboxItem.user_id == user_id))
        .values(read=True)
        .returning(InboxItem.id)
    )
    return session.execute(statement).scalar_one_or_none() is not None


def create_inbox_item(
    session: Session,
    *,
    user_id: UUID,
    item_type: str,
    title: str | None,
    content: dict,
) -> InboxItem:
    item = InboxItem(user_id=user_id, item_type=item_type, title=title, content=content)
    session.add(item)
    session.flush()
    return item

File: src/lib/api/inbox.ts
import { and, desc, eq, inArray, sql } from 'drizzle-orm';
export {
  INBOX_FILTER_TABS,
  INBOX_FILTER_TO_TYPES,
  INBOX_ITEM_TYPES,
  type InboxFilter,
  type InboxItemRow,
  type InboxItemType,
} from '@/lib/inbox-shared';
import {
  INBOX_FILTER_TO_TYPES,
  type InboxFilter,
  type InboxItemRow,
  type InboxItemType,
} from '@/lib/inbox-shared';

export async function listInboxItemsByUserId(
  userId: string,
  filter: InboxFilter = 'all'
): Promise<InboxItemRow[]> {
  const [{ db }, { inboxItems }] = await Promise.all([
    import('@/lib/db'),
    import('@/lib/db/schema'),
  ]);

  const selectedTypes = INBOX_FILTER_TO_TYPES[filter];
  const condition =
    selectedTypes === null
      ? eq(inboxItems.userId, userId)
      : and(
          eq(inboxItems.userId, userId),
          inArray(inboxItems.itemType, selectedTypes)
        );

  const rows = await db
    .select({
      id: inboxItems.id,
      itemType: inboxItems.itemType,
      title: inboxItems.title,
      content: inboxItems.content,
      read: inboxItems.read,
      createdAt: inboxItems.createdAt,
    })
    .from(inboxItems)
    .where(condition)
    .orderBy(desc(inboxItems.createdAt));

  return rows.map((row) => ({
    id: row.id,
    itemType: row.itemType as InboxItemType,
    title: row.title ?? '',
    content: (row.content as Record<string, unknown>) ?? {},
    read: row.read,
    createdAt: row.createdAt.toISOString(),
  }));
}

export async function markInboxItemRead(
  id: string,
  userId: string
): Promise<boolean> {
  const [{ db }, { inboxItems }] = await Promise.all([
    import('@/lib/db'),
    import('@/lib/db/schema'),
  ]);

  const rows = await db
    .update(inboxItems)
    .set({ read: true })
    .where(and(eq(inboxItems.id, id), eq(inboxItems.userId, userId)))
    .returning({ id: inboxItems.id });

  return rows.length > 0;
}

export async function getUnreadInboxCount(userId: string): Promise<number> {
  const [{ db }, { inboxItems }] = await Promise.all([
    import('@/lib/db'),
    import('@/lib/db/schema'),
  ]);

  const rows = await db
    .select({ count: sql<number>`count(*)::int` })
    .from(inboxItems)
    .where(and(eq(inboxItems.userId, userId), eq(inboxItems.read, false)));

  return rows[0]?.count ?? 0;
}

export async function createInboxItem(input: {
  userId: string;
  itemType: InboxItemType;
  title: string;
  content: Record<string, unknown>;
  read?: boolean;
}): Promise<string> {
  const [{ db }, { inboxItems }] = await Promise.all([
    import('@/lib/db'),
    import('@/lib/db/schema'),
  ]);

  const rows = await db
    .insert(inboxItems)
    .values({
      userId: input.userId,
      itemType: input.itemType,
      title: input.title,
      content: input.content,
      read: input.read ?? false,
    })
    .returning({ id: inboxItems.id });

  return rows[0]?.id ?? '';
}
